src/libcollections/btree/node.rs

```diff
@@ -1323,57 +1333,127 @@ impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Internal>, marker::
+            let (right_len, right_k, right_v, right_e) = self.reborrow_mut()
+                .right_edge().descend().into_pointers_mut();
+
+            // Get raw pointers to parent's key and value.
+            let (parent_k, parent_v) = {
+                let kv = self.reborrow_mut().into_kv_mut();
+                (kv.0 as *mut K, kv.1 as *mut V)
+            };
+
+            // Make sure that we may steal safely.
+            debug_assert!(left_len + n <= CAPACITY);
+            debug_assert!(right_len >= n);
+
+            // Move parent key/value pair to the left child.
+            move_kv(parent_k, parent_v, 0, left_k, left_v, left_len, 1);
+
+            // Move elements from the right child to the left one.
+            move_kv(right_k, right_v, 0, left_k, left_v, left_len + 1, n - 1);
+            move_edges(right_e, 0, left_e, left_len + 1, n);
+
+            // Move right-most stolen pair to parent.
+            move_kv(right_k, right_v, n - 1, parent_k, parent_v, 0, 1);
+
+            // Fix right indexing
+            let new_right_len = right_len - n;
+            ptr::copy(right_k.offset(n as isize),
+                      right_k,
+                      new_right_len);
+            ptr::copy(right_v.offset(n as isize),
+                      right_v,
+                      new_right_len);
+            if let Some(edges) = right_e {
+                ptr::copy(edges.offset(n as isize),
+                          edges,
+                          new_right_len + 1);
+            }
+
+            // Fix left length and childs' parent information.
+            {
+                let mut left = self.reborrow_mut().left_edge().descend();
+                left.as_leaf_mut().len += n as u16;
+                if let ForceResult::Internal(mut node) = left.force() {
+                    node.correct_childs_parent_links();
+                }
+            }
+
+            // Fix right length and childs' parent information.
+            {
+                let mut right = self.reborrow_mut().right_edge().descend();
+                right.as_leaf_mut().len -= n as u16;
+                if let ForceResult::Internal(mut node) = right.force() {
+                    node.correct_childs_parent_links();
                 }
             }
         }
     }
 }
 
+unsafe fn move_kv<K, V>(
+    source_k: *mut K, source_v: *mut V, source_offset: usize,
+    dest_k: *mut K, dest_v: *mut V, dest_offset: usize,
+    count: usize)
+{
+    ptr::copy_nonoverlapping(source_k.offset(source_offset as isize),
+                             dest_k.offset(dest_offset as isize),
+                             count);
+    ptr::copy_nonoverlapping(source_v.offset(source_offset as isize),
+                             dest_v.offset(dest_offset as isize),
+                             count);
+}
+
+// Source and destination must have the same height.
+unsafe fn move_edges<K, V>(
+    source: Option<*mut BoxedNode<K, V>>, source_offset: usize,
```

Comment:
I think I'd prefer if this function took pointers not wrapped in `Option` - it makes more sense intuitively to me that you would check if you have edges to move, and if so, you call `move_edges`. I could be convinced otherwise though.